# Het super sociaal netwerk

In deze notebook verken je verschillende technieken om inzicht te krijgen in een sociaal netwerk. Je zal bekijken hoe uitgespreid het netwerk is. Welke influencers er zijn in het netwerk en welke communities er zijn in het netwerk. 

Voor je start, installeer en importeer je de nodige bibliotheken.

In [46]:
!pip install networkx

In [47]:
import csv
import numpy as np
import networkx as nx

## Het sociaal netwerk laden

In de map *dataset* kan je een bestand vinden met de naam *hero-network.csv*. Elke lijn van dit bestand bevat twee namen van superhelden. Zo'n paar werd aan het bestand toegevoegd wanneer de twee helden samen voorkomen in een strip. 

**Opdracht**: 
- Open het bestand *hero-network.csv* in de map *dataset*;
- Bekijk het bestand.
- Zijn er helden die je herkent? 
- Zijn er helden die je niet herkent?

Met de volgende codecel laden we de inhoud van het csv bestand in in een numpy array.

In [66]:
# Lees het csv bestand.
data = None
with open('./dataset/hero-network-small-sample.csv', newline='') as csvfile:
    data = np.array(list(csv.reader(csvfile)))
    
# Verwijder de eerste rij, deze bevat de kolomnamen.
data = data[1:]

## Inzicht krijgen in de dataset

Voor je start met een analyse op een dataset, is het goed om een idee te krijgen van de gegevens die in die dataset zitten. Hieronder verkennen we een aantal eenvoudige eigenschappen van de dataset.

Voer de volgende codecel uit om 20 willekeurige heldenkoppels af te drukken.

In [67]:
# Print 20 radom rijen van de data.
print(data[np.random.randint(0, len(data), 20)])

[['DRAGONFLY II/MEIKO Y' 'WATSON-PARKER, MARY ']
 ["SHANNA/SHANNA O'HARA" 'VASHTI']
 ['BALDER [ASGARDIAN]' 'MASTERSON, KEVIN']
 ['AUSTIN, SUSAN' 'HOGUN [ASGARDIAN]']
 ['THING/BENJAMIN J. GR' 'WATSON-PARKER, MARY ']
 ['CYCLOPS/SCOTT SUMMER' 'KINCAID, DR. JANE FO']
 ['BLACK WIDOW/NATASHA ' 'STANKOWICZ, FABIAN']
 ['ANGEL/WARREN KENNETH' 'SHIP']
 ['INVISIBLE WOMAN/SUE ' 'PAIBOK, CAPT. [SKRUL']
 ['CYCLOPS/SCOTT SUMMER' 'ICEMAN/ROBERT BOBBY ']
 ['CALLISTO' 'MARROW/SARAH ']
 ['BISHOP /' 'WOLVERINE/LOGAN ']
 ['CAPTAIN BRITAIN/BRIA' 'ICEMAN/ROBERT BOBBY ']
 ['JAMESON, J. JONAH' 'MAYE, APRIL']
 ['MASTERSON, KEVIN' 'STEELE, MARCY MASTER']
 ['COUNTERWEIGHT/JACK P' 'NIGHTCRAWLER/KURT WA']
 ['HYDROMAN/MORRIE BENC' 'RHINO']
 ['THRENODY' 'X-MAN/NATHAN GREY ']
 ['CAPTAIN AMERICA' 'IKARIS/IKE HARRIS [E']
 ['GEE/ALEX POWER' 'JUSTICE II/VANCE AST']]


Door het formaat van onze numpy array af te drukken, komen we te weten hoeveel heldenkoppels er in de dataset zitten.

In [68]:
# Druk het formaat af van de dataset.
print(f"Er zitten {data.shape[0]} duo's in de dataset.")

Er zitten 5542 duo's in de dataset.


Wanneer je het csv bestand bekijkt, zal je zien dat er veel helden meerdere keren in de dataset voorkomen. Met de volgende codecel tellen we hoeveel unieke helden er in de dataset voorkomen.

In [69]:
# Verzamel de unieke helden in een lijst.
helden = np.unique(data)
print(f"Er zitten {len(helden)} unieke helden in de dataset.")

Er zitten 2220 unieke helden in de dataset.


We kunnen de helden ook alfabetisch afdrukken.

In [70]:
helden_alfabetisch = list(np.sort(helden))
for held in helden_alfabetisch[:20]:
    print(f"{held}")

4-D MAN/MERCURIO
A'YIN
ABEL
ABSORBING MAN/CARL C
ACHEBE, REVEREND DOC
ACHILLES II/HELMUT
ADAM X
ADAMS, CINDY
ADAMS, NICOLE NIKKI
ADORA
ADVA
ADVERSARY
AENTAROS
AGAMEMNON III/
AGED GENGHIS
AGEE, DR. AUBREY
AGONY/
AHAB/RORY CAMPBELL
AHURA
AINET


Om te weten in hoeveel strips twee helden met elkaar in contact komen, tellen we het aantal keer dat elk koppel voorkomt in de dataset. Merk op dat de volgorde waarin de namen van de helden in de dataset staan niet van belang is. Het koppel (Captain America, The Hulk) is hetzelfde als (The Hulk, Captain Amerika).

In [71]:
# Maak een dictionary die het aantal interacties per koppel helden bijhoudt.

# Initialiseer de dictionary.
interacties = {}

# Overloop alle heldenkoppels
for rij in data:
    # Sorteer de namen van het koppel zodat de volgorde niet uitmaakt.
    heldenpaar = tuple(sorted(rij))
    # Voeg het heldenpaar toe aan de dictionary of verhoog het aantal interacties.
    if heldenpaar in interacties:
        interacties[heldenpaar] += 1
    else:
        interacties[heldenpaar] = 1
        

Hieronder drukken we af hoe vaak de eerste 10 heldenparen voorkomen in de dataset.

In [72]:
for paar in list(interacties.keys())[:20]:
    print(f"{paar[0]} en {paar[1]} interageren {interacties[paar]} keer.")

4-D MAN/MERCURIO en FURY, COL. NICHOLAS  interageren 1 keer.
A'YIN en CAPTAIN BRITAIN/BRIA interageren 1 keer.
A'YIN en NIGHTCRAWLER/KURT WA interageren 1 keer.
ABEL en CANNONBALL II/SAM GU interageren 1 keer.
ABSORBING MAN/CARL C en ROBERTSON, MARTHA interageren 1 keer.
ABSORBING MAN/CARL C en MURDOCK, BATTLING JA interageren 1 keer.
ABSORBING MAN/CARL C en VIZIER interageren 1 keer.
ABSORBING MAN/CARL C en HERCULES [GREEK GOD] interageren 1 keer.
ABSORBING MAN/CARL C en WASP/JANET VAN DYNE  interageren 2 keer.
ABSORBING MAN/CARL C en PHARAOH RAMA-TUT interageren 1 keer.
ABSORBING MAN/CARL C en GALACTUS/GALAN interageren 1 keer.
ABSORBING MAN/CARL C en JAGRFELM [ASGARDIAN] interageren 1 keer.
ABSORBING MAN/CARL C en QUASAR III/WENDELL V interageren 1 keer.
ABSORBING MAN/CARL C en LIZARD/DR. CURTIS CO interageren 1 keer.
ABSORBING MAN/CARL C en BALDER [ASGARDIAN] interageren 1 keer.
ABSORBING MAN/CARL C en SHE-HULK/JENNIFER WA interageren 1 keer.
ABSORBING MAN/CARL C en ENCHANTRESS/AMO

## Een graaf opstellen

Om een graaf te bouwen maken we gebruik van de networkx bibliotheek. Deze bibliotheek maakt het gemakkelijk om analyses te doen op grafen.

Eerst en vooral maken we een graafobject aan. Aan dit object gaan we onze knopen en bogen toevoegen.

In [73]:
# Maak een graafobject aan.
graaf = nx.Graph()

We kunnen op een eenvoudige manier onze helden toevoegen als knopen aan de graaf. Daarvoor gebruiken we de functie `add_nodes_from`. We geven een lijst mee aan deze functie. De functie zal dan voor elk element uit de lijst een knoop toevoegen.

In [74]:
# Voeg de helden toe als knopen.
graaf.add_nodes_from(helden)

Nu moeten we enkel nog de bogen toevoegen aan de graaf. Daarvoor kunnen we onze dictionary met interacties gebruiken. Elke interactie komt overeen met een boog tussen de knopen van de twee helden in het paar. Het gewicht van de boog is het aantal interacties die dat paar heeft.

In [75]:
# Voeg de interacties toe als gewogen bogen.
for paar in interacties:
    graaf.add_edge(paar[0], paar[1], weight=interacties[paar])

## Influencers opsporen

Zoals je eerder in het leerpad zag, kan je belangrijke mensen in een sociaal netwerk opsporen aan de hand van hun **graadcentraliteit**. Networkx heeft een eenvoudige methode om de graadcentraliteit van elke knoop te bepalen. Onderstaande code drukt de 10 helden af met hoogste graadcentraliteit.

In [76]:
graadcentraliteit = nx.degree_centrality(graaf)
gesorteerde_graadcentraliteit = sorted(graadcentraliteit.items(), key=lambda x: x[1], reverse=True)

for held, graad in gesorteerde_graadcentraliteit[:10]:
    print(f"{held} heeft een graadcentraliteit van {graad}.")

CAPTAIN AMERICA heeft een graadcentraliteit van 0.050473186119873815.
SPIDER-MAN/PETER PAR heeft een graadcentraliteit van 0.04281207751239297.
IRON MAN/TONY STARK  heeft een graadcentraliteit van 0.041910770617395225.
WOLVERINE/LOGAN  heeft een graadcentraliteit van 0.039206849932401984.
THING/BENJAMIN J. GR heeft een graadcentraliteit van 0.035601622352410996.
THOR/DR. DONALD BLAK heeft een graadcentraliteit van 0.03334835511491663.
VISION  heeft een graadcentraliteit van 0.03334835511491663.
INVISIBLE WOMAN/SUE  heeft een graadcentraliteit van 0.03199639477242001.
PROFESSOR X/CHARLES  heeft een graadcentraliteit van 0.031545741324921134.
CYCLOPS/SCOTT SUMMER heeft een graadcentraliteit van 0.03064443442992339.


**Opdracht**: Naast graadcentraliteit kan je ook de eigenwaardecentraliteit bepalen met de functie `eigenvector_centrality`. Gebruik onderstaande codecel om de 10 populairste helden af te drukken volgens eigenwaardecentraliteit.

In [77]:
eigenwaardecentraliteit = nx.eigenvector_centrality(graaf)
gesorteerde_eigenwaardecentraliteit = sorted(eigenwaardecentraliteit.items(), key=lambda x: x[1], reverse=True)

for held, graad in gesorteerde_eigenwaardecentraliteit [:10]:
    print(f"{held} heeft een eigenwaardecentraliteit van {graad}.")

CAPTAIN AMERICA heeft een eigenwaardecentraliteit van 0.22623967278691517.
THING/BENJAMIN J. GR heeft een eigenwaardecentraliteit van 0.20405164905465953.
THOR/DR. DONALD BLAK heeft een eigenwaardecentraliteit van 0.1847508966981997.
SPIDER-MAN/PETER PAR heeft een eigenwaardecentraliteit van 0.1769924765281653.
VISION  heeft een eigenwaardecentraliteit van 0.17352137811323107.
IRON MAN/TONY STARK  heeft een eigenwaardecentraliteit van 0.16980993569018116.
SCARLET WITCH/WANDA  heeft een eigenwaardecentraliteit van 0.16792818393718284.
WOLVERINE/LOGAN  heeft een eigenwaardecentraliteit van 0.1584441117403304.
MR. FANTASTIC/REED R heeft een eigenwaardecentraliteit van 0.15749802704553423.
ANT-MAN/DR. HENRY J. heeft een eigenwaardecentraliteit van 0.15694266580050986.


**Opdracht**: Vergelijk de populairste helden volgens eigenwaarde- en graadcentraliteit. Welke verschillen zie je?

## Spreiding van het netwerk

De spreiding van het netwerk is de maximale kortste afstand tussen elke twee helden in het netwerk. Voor elk paar van knopen berekenen we met het algoritme van Dijkstra de lengte van het kortste pad tussen deze twee knopen. Van al deze afstanden nemen we de langste afstand als *diameter* van het netwerk. Deze diameter zegt ons iets over hoe uitgespreid het netwerk is. Hoe kleiner de diameter, hoe sterker de knopen in het netwerk met elkaar verbonden zijn.




**Merk op dat de code in de volgende cel veel tijd in beslag kan nemen. De computer moet immers de afstand berekenen tussen elk paar knopen in de graaf er zijn zo'n 2220 x 2220 = 4928400 mogelijke paren.**

In [83]:
# Calculate the maximal shortest path between any two nodes
diameter = nx.diameter(graaf)
print(f"The diameter van de graaf is {diameter}")

The diameter van de graaf is 13


## Communities zoeken

Naast de influencers in het netwerk en de spreiding van het netwerk, kunnen we ook op zoek gaan naar communities binnen het netwerk. In een sociaal netwerk komt een community vaak overeen met een specifieke vriendengroep. Bijvoorbeeld mensen die elkaar kennen van in de sportclub.

Om communities te detecteren maken we gebruik van de Louvain methode. Dit is een algoritme die stapsgewijs op zoek zal gaan naar communities binnen het netwerk. Dit doet het algoritme door de *modulariteit* van het netwerk te berekenen en deze stap voor stap te maximaliseren. Hier gebruiken we een speciale bibliotheek die al een implementatie van het algoritme bevat.

Eerst installeren we de bibliotheek.

In [84]:
!pip install python-louvain

  Using cached python_louvain-0.16-py3-none-any.whl


Dan importeren we de bibliotheek.

In [85]:
import community

Met deze bibliotheek kunnen we op een eenvoudige manier op zoek gaan naar "vriendengroepen" in ons netwerk.

In [86]:
vriendengroepen = community.best_partition(graaf)

Eerst en vooral bekijken we hoeveel vriendengroepen het algoritme gevonden heeft.

In [89]:
# Druk het aantal vriendengroepen in het netwerk af.
print(f"Er zijn {len(set(vriendengroepen.values()))} vriendengroepen in het netwerk.")

Er zijn 35 vriendengroepen in het netwerk.


We kunnen ook bekijken hoeveel helden in elke vriendengroep zitten.

In [90]:
# Druk voor elke vriendengroep het aantal leden af.
for vriendengroep in set(vriendengroepen.values()):
    leden = [held for held, groep in vriendengroepen.items() if groep == vriendengroep]
    print(f"Vriendengroep {vriendengroep} heeft {len(leden)} leden.")

Vriendengroep 0 heeft 85 leden.
Vriendengroep 1 heeft 463 leden.
Vriendengroep 2 heeft 349 leden.
Vriendengroep 3 heeft 216 leden.
Vriendengroep 4 heeft 59 leden.
Vriendengroep 5 heeft 8 leden.
Vriendengroep 6 heeft 148 leden.
Vriendengroep 7 heeft 141 leden.
Vriendengroep 8 heeft 5 leden.
Vriendengroep 9 heeft 76 leden.
Vriendengroep 10 heeft 82 leden.
Vriendengroep 11 heeft 5 leden.
Vriendengroep 12 heeft 41 leden.
Vriendengroep 13 heeft 100 leden.
Vriendengroep 14 heeft 71 leden.
Vriendengroep 15 heeft 25 leden.
Vriendengroep 16 heeft 20 leden.
Vriendengroep 17 heeft 8 leden.
Vriendengroep 18 heeft 4 leden.
Vriendengroep 19 heeft 3 leden.
Vriendengroep 20 heeft 114 leden.
Vriendengroep 21 heeft 18 leden.
Vriendengroep 22 heeft 42 leden.
Vriendengroep 23 heeft 9 leden.
Vriendengroep 24 heeft 3 leden.
Vriendengroep 25 heeft 41 leden.
Vriendengroep 26 heeft 6 leden.
Vriendengroep 27 heeft 16 leden.
Vriendengroep 28 heeft 16 leden.
Vriendengroep 29 heeft 18 leden.
Vriendengroep 30 heeft

Om de populairste helden in elke vriendengroep te vinden, kunnen we de helden binnen een vriendengroep rangschikken volgens graadcentraliteit. Onderstaande cel drukt voor elke vriendengroep de 5 populairste helden af.

In [92]:
# Druk de eerste 5 leden van elke vriendengroep af gerangschikt volgens graadcentraliteit.
for vriendengroep in set(vriendengroepen.values()):
    leden = [held for held, groep in vriendengroepen.items() if groep == vriendengroep]
    graadcentraliteit_van_leden = {held: graadcentraliteit[held] for held in leden}
    gesorteerde_leden = sorted(graadcentraliteit_van_leden.items(), key=lambda x: x[1], reverse=True)
    print(f"Vriendengroep {vriendengroep} heeft de volgende top 5 leden:")
    for held, graad in gesorteerde_leden[:5]:
        print(f"    - {held} met een graadcentraliteit van {graad}.")

Vriendengroep 0 heeft de volgende top 5 leden:
    - FURY, COL. NICHOLAS  met een graadcentraliteit van 0.018026137899954935.
    - MOON KNIGHT/MARC SPE met een graadcentraliteit van 0.008562415502478594.
    - DUGAN, TIMOTHY ALOYI met een graadcentraliteit van 0.005858494817485354.
    - DEATH met een graadcentraliteit van 0.005407841369986481.
    - PUNISHER II/FRANK CA met een graadcentraliteit van 0.005407841369986481.
Vriendengroep 1 heeft de volgende top 5 leden:
    - WOLVERINE/LOGAN  met een graadcentraliteit van 0.039206849932401984.
    - PROFESSOR X/CHARLES  met een graadcentraliteit van 0.031545741324921134.
    - CYCLOPS/SCOTT SUMMER met een graadcentraliteit van 0.03064443442992339.
    - BEAST/HENRY &HANK& P met een graadcentraliteit van 0.02794051374493015.
    - STORM/ORORO MUNROE S met een graadcentraliteit van 0.027039206849932402.
Vriendengroep 2 heeft de volgende top 5 leden:
    - CAPTAIN AMERICA met een graadcentraliteit van 0.050473186119873815.
    - IRON MAN/T

**Opdracht:** Bekijk de uitvoer van bovenstaande cel. Hoe worden de vriendengroepen volgens jou gegroepeerd? Begrijp je waarom deze helden in dezelfde vriendengroep zitten?

# Partners

Dit materiaal werd ontwikkeld door Dwengo vzw en kwam tot stand met steun van VLAIO. Vind al het materiaal van ons wAIsda project op [dwengo.org/waisda](dwengo.org/waisda)

!["VLAIO logo"](img/vlaio.png)

!["Dwengo logo"](img/dwengo-groen-zwart.png)